### PR TITLE
Implement Outlets

### PR DESCRIPTION
**Pull request by aleemont1**

### CODE REVIEW

**Comment by aleemont1** [comment]
Okie dokie, looks better now.

File: app/src/main/java/homer/model/outlets/Outlet.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    public void plug(final Device<?> device) {
+        this.device = Optional.of(Objects.requireNonNull(device));
+    }
+
+    /**
+     * Unplug the device.
+     */
+    public void unplug() {
+        Objects.requireNonNull(this.device);
+        this.device = null;
```
Comment (smagnaniuni) [inline app/src/main/java/homer/model/outlets/Outlet.java:112]:
```suggestion
        this.device = Optional.empty();
```

File: app/src/main/java/homer/model/outlets/Outlet.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        this.state = value;
+    }
+
+    /**
+     * Plugs a device to the outlet.
+     * 
+     * @param device The {@link homer.api.Device} to plug.
+     */
+    public void plug(final Device<?> device) {
+        this.device = Optional.of(Objects.requireNonNull(device));
```
Comment (smagnaniuni) [inline app/src/main/java/homer/model/outlets/Outlet.java:104]: Optional.of() will throw a NullPointerException if device is null here, so maybe `Objects.requireNonNull` is redundant. Maybe I'll ask @michelececcacci about this

**Review by smagnaniuni — COMMENTED** [review]
See the comments but lgtm

File: app/src/main/java/homer/model/outlets/Outlet.java (hunk 2 above)
Comment (aleemont1) [reply app/src/main/java/homer/model/outlets/Outlet.java:104]: From prof. Viroli's [slides](https://virtuale.unibo.it/mod/resource/view.php?id=1081730) (Look on slide 22).
```
public void setPartner (final Person p) {
   this.partner = Optional.of(Objects.requireNonNull(p));
}
```

File: app/src/main/java/homer/model/outlets/Outlet.java (hunk 1 above)
Comment (aleemont1) [reply app/src/main/java/homer/model/outlets/Outlet.java]: Fixed.

File: app/src/main/java/homer/model/outlets/Outlet.java (hunk 2 above)
Comment (smagnaniuni) [reply app/src/main/java/homer/model/outlets/Outlet.java:104]: Pretty sure it's an error, [Optional.of](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Optional.html#of(T)) throws a NullPointerException, and [requireNonNull](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Objects.html#requireNonNull(T)) too

In the openjdk implementation Optional.of uses requireNonNull itself
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/Optional.java#L113

We could write on the forum to report this

Comment (aleemont1) [reply app/src/main/java/homer/model/outlets/Outlet.java:104]: Yep, looks like you're right. We can report this then.

Comment (aleemont1) [reply app/src/main/java/homer/model/outlets/Outlet.java:104]: I'll change this tomorrow in the morining and then we can merge, right?

@michelececcacci what do you think?

**Comment by michelececcacci** [comment]
LGTM

**Pull request merged**